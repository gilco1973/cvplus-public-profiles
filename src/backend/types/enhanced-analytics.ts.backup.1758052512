// @ts-ignore
/**
 * Enhanced Analytics Types
 * 
 * Analytics and tracking types for enhanced CV features.
 * Extracted from enhanced-models.ts to maintain <200 line compliance.
 * 
 * @author Gil Klainert
 * @version 1.0.0
 */

/**
 * Public CV profile data
 */
export interface PublicCVProfile {
  id: string;
  userId: string;
  jobId: string;
  slug: string;
  isPublic: boolean;
  seoTitle?: string;
  seoDescription?: string;
  metaTags?: Record<string, string>;
  customDomain?: string;
  publicUrl?: string;
  allowContactForm?: boolean;
  showAnalytics?: boolean;
  contactEmail?: string;
  parsedCV?: any; // Parsed CV data with PII masking
  features?: Record<string, any>; // Public profile features
  template?: string; // Template name
  qrCodeUrl?: string; // QR code URL for the profile
  analytics: PublicProfileAnalytics;
  socialSharing: {
    enabled: boolean;
    platforms: string[];
    customMessage?: string;
  };
  privacySettings?: {
    enabled?: boolean;
    level?: 'public' | 'private' | 'restricted';
    maskingRules?: {
      maskEmail?: boolean;
      maskPhone?: boolean;
      maskAddress?: boolean;
      name?: boolean;
    };
    publicEmail?: boolean;
    publicPhone?: boolean;
  };
  additionalInfo?: {
    availabilityCalendar?: any;
    testimonials?: any[];
    personalityProfile?: any;
    skillsVisualization?: any;
    certifications?: any[];
    contactEmail?: string;
  };
  createdAt: Date;
  updatedAt: Date;
  lastViewedAt?: Date;
}

/**
 * Public CV Profile analytics (summary level)
 */
export interface PublicProfileAnalytics {
  totalViews: number;
  uniqueVisitors: number;
  averageTimeOnPage: number;
  bounceRate: number;
  featureUsage: Record<string, number>;
  conversionRate: number;
  lastAnalyticsUpdate: Date;
  views?: number; // Alias for totalViews for backward compatibility
  qrScans?: number; // QR code scan count
  contactSubmissions?: number; // Contact form submissions
  lastViewedAt?: Date | null; // Last time the profile was viewed
}

/**
 * Feature analytics tracking (detailed interactions)
 */
export interface FeatureAnalytics {
  jobId: string;
  featureId: string;
  userId?: string; // Visitor ID if available
  interactions: FeatureInteraction[];
  aggregates: {
    totalInteractions: number;
    uniqueUsers: number;
    averageEngagementTime: number;
    lastInteraction: Date;
  };
  // Legacy properties for backward compatibility
  totalViews?: number;
  uniqueVisitors?: number;
  averageTimeOnPage?: number;
  bounceRate?: number;
  featureUsage?: Record<string, number>;
  conversionRate?: number;
  lastAnalyticsUpdate?: Date;
  views?: number;
  qrScans?: number;
  contactSubmissions?: number;
}

/**
 * Feature interaction tracking
 */
export interface FeatureInteraction {
  type: string; // 'view', 'click', 'submit', etc.
  timestamp: Date;
  duration?: number; // For time-based interactions
  metadata?: Record<string, any>;
  userAgent?: string;
  ipHash?: string; // Hashed IP for privacy
  // Legacy properties for backward compatibility
  featureId?: string;
  userId?: string;
  jobId?: string;
  interactionType?: 'view' | 'click' | 'download' | 'share' | 'contact';
}

/**
 * Contact form submission data
 */
export interface ContactFormSubmission {
  id: string;
  jobId: string;
  name: string;
  email: string;
  subject?: string;
  message: string;
  phoneNumber?: string;
  company?: string;
  linkedinUrl?: string;
  interestedServices?: string[];
  preferredContactMethod: 'email' | 'phone' | 'linkedin';
  isRead: boolean;
  isReplied: boolean;
  submittedAt: Date;
  ipAddress?: string;
  userAgent?: string;
  source: 'cv' | 'qr' | 'direct';
  timestamp?: Date; // For backward compatibility
  status?: string; // For status tracking
}

/**
 * QR code scan tracking
 */
export interface QRCodeScan {
  id: string;
  jobId: string;
  qrType: 'primary' | 'contact' | 'chat' | 'menu';
  scannedAt: Date;
  ipAddress?: string;
  userAgent?: string;
  location?: {
    country: string;
    city: string;
    coordinates?: {
      lat: number;
      lng: number;
    };
  };
  device: {
    type: 'mobile' | 'tablet' | 'desktop';
    os?: string;
    browser?: string;
  };
  converted: boolean;
  conversionType?: 'view' | 'contact' | 'chat' | 'download';
  scanId?: string; // For backward compatibility
  timestamp?: Date; // For backward compatibility
}