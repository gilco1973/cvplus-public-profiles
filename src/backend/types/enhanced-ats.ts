// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts/**
 * Enhanced ATS Optimization Types
 * 
 * ATS (Applicant Tracking System) optimization and analysis types.
 * Extracted from enhanced-models.ts to maintain <200 line compliance.
 * 
 * @author Gil Klainert
 * @version 1.0.0
 */

import { ParsedCV } from './job';

/**
 * Advanced ATS compatibility score
 */
export interface AdvancedATSScore {
  overall: number;
  breakdown: {
    parsing: number;
    formatting: number;
    keywords: number;
    content: number;
    specificity: number;
    experience: number;
    education: number;
    skills: number;
    achievements: number;
  };
  recommendations: PrioritizedRecommendation[];
  competitorAnalysis: CompetitorAnalysis;
  semanticKeywords: SemanticKeywordAnalysis;
  industryBenchmark: number;
  estimatedPassRate: number;
  simulationResults: ATSSystemSimulation[];
}

/**
 * Prioritized improvement recommendation
 */
export interface PrioritizedRecommendation {
  id: string;
  category: string;
  title: string;
  description: string;
  priority: 'critical' | 'high' | 'medium' | 'low';
  impact: number;
  effort: 'low' | 'medium' | 'high';
  timeEstimate: string;
  implementation: string[];
  examples?: string[];
  beforeAfter?: {
    before: string;
    after: string;
  };
  // Additional optional properties for compatibility
  estimatedScoreImprovement?: number;
  actionRequired?: 'add' | 'modify' | 'remove' | 'replace' | 'reformat';
  section?: string;
  keywords?: string[];
  atsSystemsAffected?: string[];
  currentValue?: string;
  targetValue?: string;
  timeToImplement?: string;
}

/**
 * Competitor analysis data
 */
export interface CompetitorAnalysis {
  similarProfiles: number;
  keywordGaps: string[];
  strengthsVsCompetitors: string[];
  improvementOpportunities: string[];
  marketPositioning: string;
  competitiveAdvantage: string[];
  benchmarkScore: number;
}

/**
 * Semantic keyword analysis
 */
export interface SemanticKeywordAnalysis {
  primaryKeywords: KeywordMatch[];
  secondaryKeywords: KeywordMatch[];
  missingKeywords: string[];
  keywordDensity: Record<string, number>;
  synonyms: Record<string, string[]>;
  contextualRelevance: Record<string, number>;
  industryTerms: string[];
  trendingKeywords: string[];
}

/**
 * Keyword match analysis
 */
export interface KeywordMatch {
  keyword: string;
  frequency: number;
  context: string[];
  importance: number;
  variations: string[];
}

/**
 * ATS system simulation results
 */
export interface ATSSystemSimulation {
  systemName: string;
  version: string;
  passRate: number;
  issues: ATSIssue[];
  suggestions: ATSSuggestion[];
  processingTime: number;
  confidence: number;
}

/**
 * ATS optimization result
 */
export interface ATSOptimizationResult {
  score: number; // 0-100 (maps to AdvancedATSScore.overall)
  overall: number; // Alias for score
  overallScore?: number; // Additional alias for compatibility
  confidence?: number; // New field, optional for backward compatibility
  passes: boolean;
  issues: ATSIssue[];
  suggestions: ATSSuggestion[];
  recommendations: string[]; // List of recommendation strings
  optimizedContent?: Partial<ParsedCV>;
  breakdown?: {
    parsing: number;
    formatting: number;
    keywords: number;
    content: number;
    specificity: number;
    experience: number;
    education: number;
    skills: number;
    achievements: number;
  }; // Score breakdown by category
  keywords: {
    found: string[];
    missing: string[];
    recommended: string[];
    density?: number; // Keyword density score
    suggestions?: string[]; // Used by ContentOptimizationService
  };
  originalScore: number;
  optimizedScore: number;
  improvement: number;
  changesApplied: string[];
  timeToOptimize: number;
  beforeAfterComparison: {
    keywordMatches: { before: number; after: number };
    formatIssues: { before: number; after: number };
    readabilityScore: { before: number; after: number };
  };
  industryAlignment: number;
  roleSpecificOptimizations: string[];
  nextSteps: string[];
  maintenanceSchedule: {
    nextReview: Date;
    frequency: 'weekly' | 'monthly' | 'quarterly';
    autoOptimization: boolean;
  };
  // New advanced fields (optional for backward compatibility)
  advancedScore?: AdvancedATSScore;
  semanticAnalysis?: SemanticKeywordAnalysis;
  systemSimulations?: ATSSystemSimulation[];
  competitorBenchmark?: CompetitorAnalysis; // Used by ATSOptimizationOrchestrator
  processingMetadata?: any; // Used by ContentOptimizationService
  verificationResults?: any; // Used by ATSOptimizationOrchestrator
}

/**
 * ATS parsing issue
 */
export interface ATSIssue {
  type: string;
  severity: 'critical' | 'warning' | 'info';
  description: string;
  location: string;
}

/**
 * ATS improvement suggestion
 */
export interface ATSSuggestion {
  category: string;
  suggestion: string;
  impact: 'high' | 'medium' | 'low';
  implementation: string;
}