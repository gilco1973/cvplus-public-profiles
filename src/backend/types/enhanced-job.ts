/**
 * Enhanced Job Models - Main Interface
 * 
 * Core enhanced job interface and related models for CV enhancement features.
 * Properly modularized to maintain <200 line compliance through separation of concerns.
 * 
 * @author Gil Klainert
 * @version 1.0.0
 */

// Import and re-export core enhanced job interface
export type { EnhancedJob } from './enhanced-job-core';

// Import and re-export media-related types
export type {
  PortfolioImage,
  CalendarSettings,
  Testimonial,
  PersonalityProfile
} from './enhanced-media';

// Import and re-export skills and certification types
export type {
  SkillsVisualization,
  SkillCategory,
  LanguageSkill,
  Certification
} from './enhanced-skills';

// Privacy settings type (kept here as it's small and job-specific)
export interface PrivacySettings {
  /** Show contact information publicly */
  showContactInfo: boolean;
  
  /** Show social media links */
  showSocialLinks: boolean;
  
  /** Allow CV download by visitors */
  allowCVDownload: boolean;
  
  /** Show analytics data to profile visitors */
  showAnalytics: boolean;
  
  /** Allow visitors to send chat messages */
  allowChatMessages: boolean;
  
  /** Make profile publicly accessible */
  publicProfile: boolean;
  
  /** Allow profile to be found in search engines */
  searchable: boolean;
  
  /** Display personality profile section */
  showPersonalityProfile: boolean;
  
  /** Display testimonials section */
  showTestimonials: boolean;
  
  /** Display portfolio section */
  showPortfolio: boolean;
  
  /** Privacy level enabled */
  enabled?: boolean;
  
  /** Masking rules for sensitive information */
  maskingRules?: {
    maskEmail?: boolean;
    maskPhone?: boolean;
    maskAddress?: boolean;
  };
  
  /** Whether to show public email */
  publicEmail?: boolean;
  
  /** Whether to show public phone */
  publicPhone?: boolean;
  
  /** Require contact form submission before allowing CV download */
  requireContactFormForCV: boolean;
}

// All enhanced job model interfaces are now properly modularized
// This provides a clean main interface while maintaining type safety
// and clear separation of concerns across logical boundaries