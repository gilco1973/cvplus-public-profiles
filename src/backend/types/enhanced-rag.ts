// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts// @ts-ignore - Export conflicts/**
 * Enhanced RAG and Chat Types
 * 
 * RAG system and chat functionality types for enhanced CV features.
 * Extracted from enhanced-models.ts to maintain <200 line compliance.
 * 
 * @author Gil Klainert
 * @version 1.0.0
 */

/**
 * User RAG profile for AI chat
 */
export interface UserRAGProfile {
  userId: string;
  jobId: string;
  vectorStoreId?: string;
  embeddingsGenerated: boolean;
  lastEmbeddingUpdate: Date;
  chunks: CVChunk[];
  chatHistory: ChatSession[];
  personalityContext: string;
  expertiseAreas: string[];
  conversationalTone: 'professional' | 'casual' | 'friendly' | 'technical';
  availableTopics: string[];
  restrictedTopics?: string[];
  customPrompts?: Record<string, string>;
  settings?: {
    maxTokens?: number;
    temperature?: number;
    responseFormat?: 'concise' | 'detailed' | 'conversational';
    language?: string;
    enablePersonalization?: boolean;
    personality?: string;
    systemPrompt?: string;
    allowedTopics?: string[];
  };
}

/**
 * CV content chunk for RAG
 */
export interface CVChunk {
  id: string;
  jobId: string;
  content: string;
  chunkType: 'personal' | 'experience' | 'skills' | 'education' | 'projects' | 'achievements';
  embedding?: number[];
  metadata: {
    section: string;
    importance: number;
    keywords: string[];
    entities: string[];
  };
  createdAt: Date;
}

/**
 * Chat session data
 */
export interface ChatSession {
  id: string;
  jobId: string;
  visitorId?: string;
  startedAt: Date;
  endedAt?: Date;
  messages: ChatMessage[];
  sessionDuration?: number;
  satisfaction?: 'positive' | 'neutral' | 'negative';
  topics: string[];
  leadGenerated: boolean;
  contactInfoShared: boolean;
  ipAddress?: string;
  userAgent?: string;
  referrer?: string;
  // Additional properties for service compatibility
  sessionId?: string; // For backward compatibility
  userId?: string; // CV owner
  createdAt?: Date; // For backward compatibility
  lastActivity?: Date; // For backward compatibility
  metadata?: {
    source?: 'public' | 'private' | 'shared';
    referrer?: string;
  };
}

/**
 * Individual chat message
 */
export interface ChatMessage {
  id: string;
  sessionId: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: Date;
  responseTime?: number;
  confidence?: number;
  sourceChunks?: string[];
  metadata?: {
    intent?: string;
    entities?: string[];
    sentiment?: 'positive' | 'neutral' | 'negative';
  };
}