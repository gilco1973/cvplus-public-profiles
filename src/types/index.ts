// Profile Types
export * from './profile.types';

// SEO Types
export * from './seo.types';

// Analytics Types
export * from './analytics.types';

// Networking Types
export * from './networking.types';

// Branding and Template Types
export interface BrandingTypes {
  colorScheme: 'light' | 'dark' | 'auto' | 'custom';
  primaryColor: string;
  secondaryColor: string;
  accentColor: string;
  fontFamily: string;
  logoUrl?: string;
}

// Social Integration Types
export interface SocialIntegrationTypes {
  platform: string;
  enabled: boolean;
  shareUrl: string;
  apiKey?: string;
  accessToken?: string;
}

// Error Types
export interface PublicProfilesError {
  code: string;
  message: string;
  details?: unknown;
  timestamp: Date;
  context?: string;
}

// API Response Types
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: PublicProfilesError;
  meta?: {
    total?: number;
    page?: number;
    limit?: number;
    hasNext?: boolean;
    hasPrevious?: boolean;
  };
}

// Configuration Types
export interface PublicProfilesConfig {
  baseUrl: string;
  apiVersion: string;
  features: {
    seoOptimization: boolean;
    analytics: boolean;
    networking: boolean;
    customDomains: boolean;
    socialIntegration: boolean;
    qrCodes: boolean;
    pdfExport: boolean;
  };
  limits: {
    maxProfiles: number;
    maxCustomDomains: number;
    maxStorageSize: number;
    maxConnections: number;
  };
}

// Event Types for Real-time Updates
export interface ProfileEvent {
  type: 'profile_updated' | 'profile_viewed' | 'connection_request' | 'message_received' | 'endorsement_received';
  profileId: string;
  userId: string;
  data: unknown;
  timestamp: Date;
}

// Utility Types
export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};

export type Optional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

export type RequiredFields<T, K extends keyof T> = T & Required<Pick<T, K>>;

export type PublicProfilesModule = 'core' | 'seo' | 'analytics' | 'networking' | 'social' | 'templates';

export interface ModuleStatus {
  module: PublicProfilesModule;
  enabled: boolean;
  version: string;
  lastUpdated: Date;
  dependencies: string[];
}