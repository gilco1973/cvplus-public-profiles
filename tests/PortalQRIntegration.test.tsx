/**
 * PortalQRIntegration.test.tsx - Unit Tests
 * 
 * Comprehensive test suite for the PortalQRIntegration component
 * 
 * @author Gil Klainert
 * @version 1.0.0
 */

import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { PortalQRIntegration } from '../PortalQRIntegration';
import { PortalConfig, QRCodeConfig } from '../../../../types/portal-types';

import { vi } from 'vitest';

// Mock Firebase functions
vi.mock('../../../../lib/firebase', () => ({
  functions: {}
}));

vi.mock('firebase/functions', () => ({
  httpsCallable: vi.fn(() => vi.fn().mockResolvedValue({ data: {} }))
}));

// Mock QR code library
vi.mock('qrcode', () => ({
  toCanvas: vi.fn().mockResolvedValue(undefined)
}));

// Mock toast notifications
vi.mock('react-hot-toast', () => ({
  success: vi.fn(),
  error: vi.fn(),
  default: {
    success: vi.fn(),
    error: vi.fn()
  }
}));

// ============================================================================
// TEST DATA
// ============================================================================

const mockPortalConfig: PortalConfig = {
  id: 'test-portal-123',
  name: 'Test Portal',
  description: 'Test portal description',
  visibility: 'public',
  theme: {
    primaryColor: '#3B82F6',
    secondaryColor: '#10B981',
    backgroundColor: '#FFFFFF',
    textColor: '#1F2937',
    fontFamily: 'Inter',
    layout: 'modern',
    animations: true,
    darkMode: false
  },
  features: {
    aiChat: true,
    qrCode: true,
    contactForm: true,
    calendar: true,
    portfolio: true,
    socialLinks: true,
    testimonials: true,
    analytics: true
  },
  metadata: {
    title: 'Test Portal',
    description: 'Test portal description',
    keywords: ['test']
  },
  createdAt: new Date(),
  updatedAt: new Date()
};

const mockQRConfig: QRCodeConfig = {
  url: 'https://test.cvplus.app',
  size: 256,
  errorCorrectionLevel: 'M',
  colors: {
    foreground: '#000000',
    background: '#FFFFFF'
  },
  includeLogo: false
};

const defaultProps = {
  portalConfig: mockPortalConfig,
  qrConfig: mockQRConfig,
  jobId: 'test-job-123',
  profileId: 'test-profile-456',
  isEnabled: true,
  mode: 'private' as const
};

// ============================================================================
// TEST SUITES
// ============================================================================

describe('PortalQRIntegration', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });
  
  // ==========================================================================
  // BASIC RENDERING TESTS
  // ==========================================================================
  
  describe('Basic Rendering', () => {
    it('renders without crashing', () => {
      render(<PortalQRIntegration {...defaultProps} />);
      expect(screen.getByRole('canvas')).toBeInTheDocument();
    });
    
    it('does not render when disabled', () => {
      const { container } = render(
        <PortalQRIntegration {...defaultProps} isEnabled={false} />
      );
      expect(container.firstChild).toBeNull();
    });
    
    it('displays title when showTitle is true', () => {
      render(
        <PortalQRIntegration
          {...defaultProps}
          displaySettings={{
            showTitle: true,
            title: 'Test QR Code Title'
          }}
        />
      );
      expect(screen.getByText('Test QR Code Title')).toBeInTheDocument();
    });
    
    it('displays instructions when showInstructions is true', () => {
      render(
        <PortalQRIntegration
          {...defaultProps}
          displaySettings={{
            showInstructions: true,
            instructions: 'Test instructions'
          }}
        />
      );
      expect(screen.getByText('Test instructions')).toBeInTheDocument();
    });
  });
  
  // ==========================================================================
  // ACTION BUTTON TESTS
  // ==========================================================================
  
  describe('Action Buttons', () => {
    it('shows download button when showDownload is true', () => {
      render(
        <PortalQRIntegration
          {...defaultProps}
          displaySettings={{ showDownload: true }}
        />
      );
      expect(screen.getByText('Download')).toBeInTheDocument();
    });
    
    it('shows share button when showShare is true', () => {
      render(
        <PortalQRIntegration
          {...defaultProps}
          displaySettings={{ showShare: true }}
        />
      );
      expect(screen.getByText('Share')).toBeInTheDocument();
    });
    
    it('opens share menu when share button is clicked', async () => {
      render(
        <PortalQRIntegration
          {...defaultProps}
          displaySettings={{ showShare: true }}
        />
      );
      
      const shareButton = screen.getByText('Share');
      fireEvent.click(shareButton);
      
      await waitFor(() => {
        expect(screen.getByText('Copy')).toBeInTheDocument();
      });
    });
    
    it('shows analytics button', () => {
      render(<PortalQRIntegration {...defaultProps} />);
      expect(screen.getByText('Analytics')).toBeInTheDocument();
    });
    
    it('shows settings button', () => {
      render(<PortalQRIntegration {...defaultProps} />);
      expect(screen.getByText('Settings')).toBeInTheDocument();
    });
  });
  
  // ==========================================================================
  // SETTINGS PANEL TESTS
  // ==========================================================================
  
  describe('Settings Panel', () => {
    it('opens settings panel when settings button is clicked', async () => {
      render(<PortalQRIntegration {...defaultProps} />);
      
      const settingsButton = screen.getByText('Settings');
      fireEvent.click(settingsButton);
      
      await waitFor(() => {
        expect(screen.getByText('QR Code Settings')).toBeInTheDocument();
      });
    });
    
    it('shows settings tabs', async () => {
      render(<PortalQRIntegration {...defaultProps} />);
      
      const settingsButton = screen.getByText('Settings');
      fireEvent.click(settingsButton);
      
      await waitFor(() => {
        expect(screen.getByText('Basic')).toBeInTheDocument();
        expect(screen.getByText('Advanced')).toBeInTheDocument();
        expect(screen.getByText('Analytics')).toBeInTheDocument();
        expect(screen.getByText('Sharing')).toBeInTheDocument();
      });
    });
    
    it('switches between settings tabs', async () => {
      render(<PortalQRIntegration {...defaultProps} />);
      
      const settingsButton = screen.getByText('Settings');
      fireEvent.click(settingsButton);
      
      await waitFor(() => {
        const advancedTab = screen.getByText('Advanced');
        fireEvent.click(advancedTab);
        
        expect(screen.getByText('Enable Logo')).toBeInTheDocument();
      });
    });
  });
  
  // ==========================================================================
  // ANALYTICS TESTS
  // ==========================================================================
  
  describe('Analytics', () => {
    it('opens analytics panel when analytics button is clicked', async () => {
      render(<PortalQRIntegration {...defaultProps} />);
      
      const analyticsButton = screen.getByText('Analytics');
      fireEvent.click(analyticsButton);
      
      await waitFor(() => {
        expect(screen.getByText('Portal QR Analytics')).toBeInTheDocument();
      });
    });
    
    it('calls onQRGenerated callback when provided', async () => {
      const mockOnQRGenerated = vi.fn();
      
      render(
        <PortalQRIntegration
          {...defaultProps}
          onQRGenerated={mockOnQRGenerated}
        />
      );
      
      // QR generation should be triggered on mount
      await waitFor(() => {
        expect(mockOnQRGenerated).toHaveBeenCalled();
      }, { timeout: 3000 });
    });
  });
  
  // ==========================================================================
  // URL GENERATION TESTS
  // ==========================================================================
  
  describe('URL Generation', () => {
    it('generates URL with UTM parameters', () => {
      const { container } = render(<PortalQRIntegration {...defaultProps} />);
      
      // Check that the final URL display contains UTM parameters
      expect(container.textContent).toContain('utm_source');
      expect(container.textContent).toContain('utm_medium');
      expect(container.textContent).toContain('utm_campaign');
    });
    
    it('includes portal ID in URL parameters', () => {
      const { container } = render(<PortalQRIntegration {...defaultProps} />);
      
      expect(container.textContent).toContain('portal_id');
    });
  });
  
  // ==========================================================================
  // ERROR HANDLING TESTS
  // ==========================================================================
  
  describe('Error Handling', () => {
    it('calls onError callback when QR generation fails', async () => {
      const mockOnError = vi.fn();
      
      // Mock QR generation to fail
      const QRCodeGenerator = await import('qrcode');
      vi.mocked(QRCodeGenerator.toCanvas).mockRejectedValueOnce(new Error('QR generation failed'));
      
      render(
        <PortalQRIntegration
          {...defaultProps}
          onError={mockOnError}
        />
      );
      
      await waitFor(() => {
        expect(mockOnError).toHaveBeenCalledWith(
          expect.objectContaining({
            code: 'QR_GENERATION_ERROR',
            message: 'QR generation failed'
          })
        );
      }, { timeout: 3000 });
    });
    
    it('displays error state when QR generation fails', async () => {
      // Mock QR generation to fail
      const QRCodeGenerator = await import('qrcode');
      vi.mocked(QRCodeGenerator.toCanvas).mockRejectedValueOnce(new Error('QR generation failed'));
      
      render(<PortalQRIntegration {...defaultProps} />);
      
      await waitFor(() => {
        expect(screen.getByText('QR Generation Error')).toBeInTheDocument();
      }, { timeout: 3000 });
    });
  });
  
  // ==========================================================================
  // CUSTOMIZATION TESTS
  // ==========================================================================
  
  describe('Customization', () => {
    it('applies custom colors from portal theme', () => {
      render(<PortalQRIntegration {...defaultProps} />);
      
      // The component should use the portal's primary color
      expect(mockPortalConfig.theme.primaryColor).toBe('#3B82F6');
    });
    
    it('applies custom size settings', async () => {
      render(
        <PortalQRIntegration
          {...defaultProps}
          displaySettings={{ size: 'large' }}
        />
      );
      
      const canvas = screen.getByRole('canvas');
      expect(canvas).toBeInTheDocument();
    });
  });
  
  // ==========================================================================
  // ACCESSIBILITY TESTS
  // ==========================================================================
  
  describe('Accessibility', () => {
    it('provides canvas element for screen readers', () => {
      render(<PortalQRIntegration {...defaultProps} />);
      
      const canvas = screen.getByRole('canvas');
      expect(canvas).toBeInTheDocument();
    });
    
    it('supports keyboard navigation for buttons', () => {
      render(
        <PortalQRIntegration
          {...defaultProps}
          displaySettings={{
            showDownload: true,
            showShare: true
          }}
        />
      );
      
      const downloadButton = screen.getByText('Download');
      const shareButton = screen.getByText('Share');
      
      expect(downloadButton).toBeInTheDocument();
      expect(shareButton).toBeInTheDocument();
      
      // Buttons should be focusable
      downloadButton.focus();
      expect(document.activeElement).toBe(downloadButton);
      
      shareButton.focus();
      expect(document.activeElement).toBe(shareButton);
    });
  });
  
  // ==========================================================================
  // INTEGRATION TESTS
  // ==========================================================================
  
  describe('Integration', () => {
    it('works with different portal configurations', () => {
      const customPortalConfig = {
        ...mockPortalConfig,
        theme: {
          ...mockPortalConfig.theme,
          primaryColor: '#FF0000',
          layout: 'creative' as const
        }
      };
      
      render(
        <PortalQRIntegration
          {...defaultProps}
          portalConfig={customPortalConfig}
        />
      );
      
      expect(screen.getByRole('canvas')).toBeInTheDocument();
    });
    
    it('handles different QR configurations', () => {
      const customQRConfig = {
        ...mockQRConfig,
        size: 512,
        errorCorrectionLevel: 'H' as const,
        includeLogo: true
      };
      
      render(
        <PortalQRIntegration
          {...defaultProps}
          qrConfig={customQRConfig}
        />
      );
      
      expect(screen.getByRole('canvas')).toBeInTheDocument();
    });
  });
});

// ============================================================================
// SNAPSHOT TESTS
// ============================================================================

describe('PortalQRIntegration Snapshots', () => {
  it('matches snapshot for basic configuration', () => {
    const { container } = render(<PortalQRIntegration {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  
  it('matches snapshot with all display options enabled', () => {
    const { container } = render(
      <PortalQRIntegration
        {...defaultProps}
        displaySettings={{
          size: 'large',
          style: 'prominent',
          showTitle: true,
          title: 'Test QR Code',
          showInstructions: true,
          instructions: 'Test instructions',
          showDownload: true,
          showShare: true
        }}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});