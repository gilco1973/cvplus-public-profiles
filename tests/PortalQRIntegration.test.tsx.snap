// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PortalQRIntegration Snapshots > matches snapshot for basic configuration 1`] = `
<div
  class="cv-feature-wrapper relative rounded-lg border transition-all duration-200 bg-gray-50 border-gray-300 shadow-sm hover:shadow-md  p-6"
>
  <div
    class="text-center"
  >
    <div
      class="w-12 h-12 mx-auto mb-4 text-red-500"
    >
      <svg
        class="w-full h-full"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.964-.833-2.732 0L3.732 16.5c-.77.833.192 2.5 1.732 2.5z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <h3
      class="text-lg font-semibold text-gray-900 mb-2"
    >
      Feature Error
    </h3>
    <p
      class="text-sm text-gray-600 mb-4"
    >
      [vitest] No "default" export is defined on the "qrcode" mock. Did you forget to return it from "vi.mock"?
If you need to partially mock a module, you can use "importOriginal" helper inside:

    </p>
    <button
      class="px-4 py-2 bg-blue-600 text-white rounded-lg hover:bg-blue-700 transition-colors"
    >
      Try Again
    </button>
  </div>
</div>
`;

exports[`PortalQRIntegration Snapshots > matches snapshot with all display options enabled 1`] = `
<div
  class="cv-feature-wrapper relative rounded-lg border transition-all duration-200 bg-gray-50 border-gray-300 shadow-sm hover:shadow-md  p-6"
>
  <div
    class="text-center"
  >
    <div
      class="w-12 h-12 mx-auto mb-4 text-red-500"
    >
      <svg
        class="w-full h-full"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.964-.833-2.732 0L3.732 16.5c-.77.833.192 2.5 1.732 2.5z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <h3
      class="text-lg font-semibold text-gray-900 mb-2"
    >
      Feature Error
    </h3>
    <p
      class="text-sm text-gray-600 mb-4"
    >
      [vitest] No "default" export is defined on the "qrcode" mock. Did you forget to return it from "vi.mock"?
If you need to partially mock a module, you can use "importOriginal" helper inside:

    </p>
    <button
      class="px-4 py-2 bg-blue-600 text-white rounded-lg hover:bg-blue-700 transition-colors"
    >
      Try Again
    </button>
  </div>
</div>
`;
